### PR TITLE
Use filelock for locking.

### PR DESCRIPTION
**Pull request by mikapfl**
As discussed in #19 , this PR switches all locking to use [filelock](https://github.com/r-lib/filelock). The main advantage is that all races should be eliminated, locks should never fail, works on cluster (gpfs), locally on linux, and on windows.

Known regressions / changes:
* An explicit queue is not kept anymore, the ordering of applicants for the lock is not guaranteed (but should still be first-in-first-out, at least on unix).
* The lock file stays in place after unlock - the presence of the lock file does not equal a locked state anymore.
* explicitly giving the folder/file on unlock is deprecated and raises a warning, all necessary information is propagated using the lock id → needs some changes in MAgPIE to remove warning.

So, to sum up, this should make locking less transparent for users (can't see the current queue, can't see if locked or not from the filesystem), but also drastically reduce the need to look at locks because they should never fail.

### CODE REVIEW

**Review by tscheypidi — APPROVED** [review]
looks good

**Comment by orichters** [comment]
Agree, this could create a mess if during the time the new gms package is loaded
- somebody starts REMIND runs 
- a REMIND run starts its subsequent run

Maybe a workflow:
- generate new library snapshots for both R versions 3 and 4
- ask everyone some days before to use them if they run longer bundles of runs or MAgPIE coupled runs
- announce a time this PR is merged and tell people not to start new runs at this moment
- ~ 10 minutes later (after `/p/projects/rd3mod/R/libraries/3.6.3/gms` and `/p/projects/rd3mod/R/libraries/4.1.2/gms` have been automatically updated), generate new library snapshots for both R versions 3 and 4 so that the latest_snapshot fallback has the new lock mode.

**Comment by tscheypidi** [comment]
I also don't see an issue here on MAgPIE side. There is an unlikely scenario that someone starts in parallel runs from the same folder but with different R sessions from which one loaded the old gms package and the other one already the new, but I would assume that the probability of such a scenario should be close to 0. At least if we select a good time for the change, e.g. at night.

But perhaps another idea to avoid this problem: What about having an intermediate gms version which contains both (old and new) implementation at the same time and which checks which one to use (old implementation, if the corresponding lock file exists, otherwise the new one). We could have that version for a few weeks before we completely switch. What do you think?

**Comment by mikapfl** [comment]
I think I like the transition plan with an intermediate version because we are not really in a hurry (sure, I would like to be done with this, but another month with an intermediate version isn't exactly a problem).

My idea:
* the filelock-locking uses `.lock2` instead of `.lock`
* the intermediate version *locks both*
* at some point in the future, we drop the traditional locking

Advantage:
* More locking never hurts, all races eliminated right away
* Time for everyone to get used to `.lock2` files and their new semantics (e.g.: they stay around)
* No need for coordinated "don't start runs now"-policies
* Lots of time to do the transition steps @orichters outlined

Disadvantage:
* Carry the old, complex locking for some more time
* Two transitions instead of one (more mental load)

**Comment by orichters** [comment]
I have to admit the "intermediate gms" suggestion sounds like unnecessary work and brain power required. If you announce today in the REMIND channel that you will merge on Sunday, this should be sufficient. Everyone with long cascades of runs in REMIND should use snapshots anyway.

**Comment by mikapfl** [comment]
Hm, I'm willing to spend two hours to make the intermediate version now, but I'm not really willing to work on Sunday to make the transition painless for REMIND if I can avoid it.

**Comment by orichters** [comment]
Fully understand it. Then say: Monday morning. I mean, if people must start runs at this time, they still can use a snapshot.

**Comment by mikapfl** [comment]
Ok, adding filelock locking to traditional locking didn't work out. Tests were failing mysteriously, and it is probably not worth diving into this deeply.

**Comment by orichters** [comment]
FYI: I created a new R4 snapshot some hours ago, needed it anyway.

**Comment by orichters** [comment]
Once the old mechanism with the `.lock` file is not used anymore, we can delete from `start.R` in REMIND the two places where `model_was_locked` is defined and used.

**Comment by mikapfl** [comment]
Ok, I added a very basic check which looks for the existence of an old lock file to the new locking code. This way, we at least catch the common case "old code locked the model, new code tries to lock it again". It just errors out, which I think is acceptable (we will tell people how to avoid this anyway, and if they mess up, a run dies instead of producing weird results).

So, the plan is now:
* I do test runs (I guess a coupled test run is the real stress test)
* I announce the date + time of the switch to the new locking and tell people to use an old snapshot when starting runs before and not to start runs during this time window (I'd say 3-4 days in advance).
* I do the switch
* I make new snapshots with the new locking directly after the switch

I hope, like this we will have minimal disruptions and can enjoy the new locking soon.

**Comment by orichters** [comment]
Sounds like a solid plan, great an end for the locking problems is in sight. It may be that I have to run a cascade of MAgPIE coupled runs in the next days/week. If you somehow manage to generate a snapshot that already contains the new locking mode, I'm willing to test that "in real life".

**Comment by mikapfl** [comment]
Interesting proposal, I'll think about it.

**Comment by orichters** [comment]
If so, please copy `/p/projects/rd3mod/R/libraries/snapshots/2022_06_09_R4` to a new folder such as `2022_06_09_newgms_R4` and replace just the `gms` subfolder, so all other libraries will remain unchanged.

**Comment by mikapfl** [comment]
> If so, please copy `/p/projects/rd3mod/R/libraries/snapshots/2022_06_09_R4` to a new folder such as `2022_06_09_newgms_R4` and replace just the `gms` subfolder, so all other libraries will remain unchanged.

Done! `2022_06_09_newgms_R4` exists and only `gms` is changed (and `filelock` added as a dependency). Worked for a `main_test.gms` test run for me, and complained about using deprecated parameters.

If you could test with a real coupling run, that would be great. I *think* it is low-risk after the tests I did, but it is possible I missed something.

Otherwise, I'll probably have to dive into the coupling documentation to find out how to start a (dummy) coupled run, that's something I haven't done so far. :sweat_smile:

**Comment by orichters** [comment]
What is your time plan? We may not start these scenarios before Wednesday. If that is too late, I can set up a REMIND-MAgPIE coupling test script within minutes for you.

**Comment by mikapfl** [comment]
For me, there is no hurry at all. Let's say, I'd like to have it done in June. So, if your scenarios finish before the end of next week, that's totally worth it for me in exchange for a thorough test.

**Comment by orichters** [comment]
first finding. Log contains warning, can be fixed [here](https://github.com/remindmodel/remind/blob/develop/scripts/start/prepare_and_run.R#L294)
```
In model_lock(timeout1 = 1, check_interval = runif(1, 10, 60)) :
  check_interval setting is deprecated and ignored.
```

**Comment by mikapfl** [comment]
There is no logging of any kind anymore when things go smoothly. Do you think we should?

**Comment by mikapfl** [comment]
And yeah, there are a couple of deprecated usages that I would remove only after the new locking becomes default.

**Comment by orichters** [comment]
> There is no logging of any kind anymore when things go smoothly. Do you think we should?

Well, it at least makes the testing a bit difficult if I can't say whether the lock actually worked. It would give me more confidence (in particular for the full set of runs) if the log states that it is waiting for unlocking as it was the case before, best with `Sys.time()`.

**Comment by mikapfl** [comment]
Hmm, it is not reliably possible to report *only* when we are waiting for a lock. We are using a blocking syscall which returns when we got the lock or when the timeout runs out. I could add logging which says "{time}: started to acquire lock" and "{time}: lock acquired" and then you can figure out yourself if the lock was granted right away or if some waiting happened. Is that reasonable?

**Comment by orichters** [comment]
Yes, that would be perfectly fine. I think this is helpful for testing now, but also later if you look at the log.txt while the model is waiting for acquiring the lock so you know why nothing is going forward at the moment.

**Comment by mikapfl** [comment]
Thinking about this further, I guess we *could* add code like this (pseudocode):
```
id = lock(timeout=1ms)
if is_timeout(id):
    print("Model already locked, waiting for lock")
    id = lock(timeout=configured_timeout)
    if is_timeout(id):
        error()
print("Acquired lock for model")
```

it would about double the complexity of the code, but would not suffer any race conditions: If we don't get the lock with the first `lock()` call, we are waiting anyway, and not holding the lock. The worst that could happen is that another process gets earlier in the queue for the lock while we are busy printing out log messaged and not waiting, but since ordering isn't guaranteed anyway…

**Comment by orichters** [comment]
I think something like this would be enough:
```
timeStartLock <- Sys.time()
message(timeStartLock, ": acquiring model lock ", appendLF = FALSE)
id = lock(timeout=configured_timeout)
timediff <- Sys.time() - timeStartLock
message("in ", round(timediff, 1), " ", units(timdiff), ".")
```

**Comment by mikapfl** [comment]
Ok, I included logging like you suggested:
```
r$> id <- gms::model_lock()
2022-06-14 16:16:39: acquiring model lock...
acquired model lock in 10.9 secs.
r$> gms::model_unlock(id)
2022-06-14 16:17:02: unlocked model.
```

**Comment by mikapfl** [comment]
Also added it to `/p/projects/rd3mod/R/libraries/snapshots/2022_06_09_newgms_R4/`

**Comment by orichters** [comment]
A minor thing: The `.lock` file has very restrictive access rights (user only). I changed it manually, but I wonder whether it is in general a good idea to give the group access to run REMIND in the same directory with different users.

**Comment by mikapfl** [comment]
Ah, very interesting indeed. Did you check if locking actually fails with default access rights when trying to lock with another user? If not, do you have time for some tests (can't test alone on this issue)?

**Comment by orichters** [comment]
Yes, the `log.txt` stated "Cannot open lock file: Permission denied". (unfortunately, I deleted the log file to restart the run with the same folder name).

For tests, feel free to use `/p/tmp/oliverr/debugging/remind_mwe` to start your runs and see them fail or let me know what I should do.

**Comment by mikapfl** [comment]
Ah, okay, then you already did the test. Thanks, good catch, I'll try to find a solution.

**Comment by mikapfl** [comment]
OK, I

* added code to create the lock file with correct permissions on unix
* rebased onto current gms
* created a new snapshot with the latest version: `/p/projects/rd3mod/R/libraries/snapshots/2022_06_09_newgms2_R4`

**Comment by orichters** [comment]
I used the `newgms` version to run the NGFS scenarios, and the results look very good. Some scenarios had to wait some minutes and I took random samples and always found another run that had locked the model. If you want to set up further tests, [here is how to do it](https://github.com/pik-piam/discussions/discussions/28).

**Comment by mikapfl** [comment]
Transition plan now:

* 2022-06-20: I make fresh snapshots for R3 and R4 with the current version of `gms` with old locking.
* 2022-06-20: I announce the transition in both RSE mattermost channels
* 2022-06-23: At 09:30-10:00, I deploy the new `gms` version with new locking to the cluster and make new snapshots for R3 and R4 with the new locking.
* :partying_face:
